pkg/hubble/metrics/metrics_test.go

```diff
@@ -55,20 +57,33 @@ func TestInitializedMetrics(t *testing.T) {
 
 }
 
+func SetUpTestMetricsServer(reg *prometheus.Registry) *httptest.Server {
+	mux := http.NewServeMux()
+	handler := promhttp.HandlerFor(reg, promhttp.HandlerOpts{
```

Comment:
This is duplicating logic in `InitMetricsServer` and now `daemon/cmd/hubble.go` is launching a Go routine for something that's already running in a Go routine. I'd rather see the `http.Server` passed into `InitMetricsServer`.

Reply:
@chancez  Sorry, I don't quite follow :smile: what would you want to see in `InitMetricsServer` and how should the product and test servers respectively be started exactly? Below are a few points if you could elaborate on the preferred solution :)

1) AFAIU you can't swap the handler on a running server instance, and I'd need pass an empty handler to create+start  the `http.Server` from a Go routine in `hubble.go` before passing it into `InitMetricsServer`, same for the test one with `httptest.NewServer`; 
2) If I pass `httptest.NewServer` into `InitMetricsServer`, it will have `ListenAndServe` called on it, which would block, unless in a Go routine, which brings us back to the timing issue in the test; it might also clash with its existing local listener from `NewUnstartedServer`.
3) I didn't add a new Go routine, just pulled the existing error-handling one up one level, out of `EnableMetrics` - previously, we were still creating these 2 Go routines sequentially: first to run `ListenAndServe`, and then to monitor the error channel - if I wrapped it again in some function the effect would be the same. Could also just move the second one (`<-errChan`)  into `InitMetricsServer`.

Reply:
Just pushed an update +resolved go.mod conflict, lmk if that's what you had in mind ;)

Reply:
Right, I see. I think your moving of handler setup to a separate function is a good improvement. 

I guess in general you cannot actually 100% correctly know if the server is started. Anything to start the server will block unless you execute it in a Go routine. Even from reading `httptest/server.go` I can't see anything that actually prevents a race condition on connecting to the server that's been started. 

There's some other issues though in your change, which I'll comment on directly.